### PR TITLE
prometheus exporter deployment

### PR DESCRIPTION
**Pull request by YanjieGao**
new exporter deployment & after test fix

### CODE REVIEW

File: service-deployment/src/gpu-exporter/dockerfile.template [hunk 1]
```diff
@@ -17,22 +17,29 @@
...
+RUN wget https://download.docker.com/linux/static/stable/x86_64/docker-17.06.2-ce.tgz
+RUN cp docker-17.06.2-ce.tgz /usr/local
+RUN tar xzvf /usr/local/docker-17.06.2-ce.tgz -C /usr/local/
+RUN cp -r /usr/local/docker/* /usr/bin/
+
+#
+# start
+#
 
+CMD python /usr/local/job_exporter.py {{ clusterconfig['clusterinfo'][ 'dataPath' ] }}/prometheus 30
```
Comment (ydye) [inline service-deployment/src/gpu-exporter/dockerfile.template]: Remove this and hardcode a path. And hardcord the path of container's fs in the kubernetes yaml file. Don't use any cluster specific configuration when building image

File: prometheus/exporter/docker_inspect.py [hunk 2]
```diff
@@ -1,9 +1,26 @@
...
+# DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
 import subprocess
 import json
 import sys
 
-targetLabel = {"PAI_HOSTNAME", "PAI_JOB_NAME", "PAI_USER_NAME", "PAI_CURRENT_TASK_ROLE_NAME", "PAI_CURRENT_GPU_ID"}
+targetLabel = {"PAI_HOSTNAME", "PAI_JOB_NAME", "PAI_USER_NAME", "PAI_CURRENT_TASK_ROLE_NAME", "GPU_ID"}
 targetEnv = {"PAI_TASK_INDEX"}
```
Comment (fanyangCS) [inline prometheus/exporter/docker_inspect.py]: >PAI_TASK_INDEX [](start = 14, length = 14)

why do you need this? we are going to retire it.

Comment (YanjieGao) [reply prometheus/exporter/docker_inspect.py]: to aggregate task level metrics

File: service-deployment/bootstrap/prometheus/node-exporter-ds.yaml.template [hunk 3]
```diff
@@ -52,26 +53,44 @@ spec:
...
         name: node-exporter
         args:
-          - '--collector.textfile.directory={{ clusterinfo[ 'dataPath' ] }}/prometheus'
+          - '--collector.textfile.directory=/datastorage/prometheus'
         ports:
         - containerPort: {{ clusterinfo['prometheusinfo']['node_exporter_port'] }}
           hostPort: {{ clusterinfo['prometheusinfo']['node_exporter_port'] }}
           name: scrape
       - image: {{ clusterinfo['dockerregistryinfo']['prefix'] }}gpu-exporter:{{ clusterinfo['dockerregistryinfo']['docker_tag'] }}
+        imagePullPolicy: Always
```
Comment (fanyangCS) [inline service-deployment/bootstrap/prometheus/node-exporter-ds.yaml.template]: >Always [](start = 25, length = 6)

why do we need this?

Comment (YanjieGao) [reply service-deployment/bootstrap/prometheus/node-exporter-ds.yaml.template]: yundong's issue https://github.com/Microsoft/pai/issues/460

Comment (YanjieGao) [reply service-deployment/bootstrap/prometheus/node-exporter-ds.yaml.template]: refresh local image

File: prometheus/exporter/job_exporter.py [hunk 4]
```diff
@@ -83,11 +88,12 @@ def genJobMetrics(logDir, gpuMetrics):
         outputFile.write(containerMemPerc)
 
 def main(argv):
+    logDir = argv[0]
+    timeSleep = int(argv[1])
     while(True):
-        logDir = argv[0]
-        timeSleep = int(argv[1])
+        print("job exporter running")
```
Comment (fanyangCS) [inline prometheus/exporter/job_exporter.py]: >print("job exporter running") [](start = 8, length = 29)

move to ln 93, before the while loop? are you going to print it in every iteration?

File: prometheus/exporter/job_exporter.py [hunk 5]
```diff
@@ -83,11 +88,12 @@ def genJobMetrics(logDir, gpuMetrics):
...
         # collect GPU metrics
-        gpuMetrics = gpu_exporter.genGpuMetricsFromSmi([logDir])
+        gpuMetrics = gpu_exporter.genGpuMetricsFromSmi(logDir)
         # join with docker stats metrics and docker inspect labels
         genJobMetrics(logDir, gpuMetrics)
```
Comment (fanyangCS) [inline prometheus/exporter/job_exporter.py]: >genJobMetrics [](start = 8, length = 13)

what is the behavior if there is no job running in this machine?

Comment (YanjieGao) [reply prometheus/exporter/job_exporter.py]: will not generate job metrics. and only generate node gpu metrics

File: prometheus/exporter/job_exporter.py (hunk 4 above)
Comment (YanjieGao) [reply prometheus/exporter/job_exporter.py]: yes hope to print evey iteration, will move

Comment (fanyangCS) [reply prometheus/exporter/job_exporter.py]: if so, suggest to add a counter or timestamp

---
In reply to: [181279874](https://github.com/Microsoft/pai/pull/472#discussion_r181279874) [](ancestors = 181279874)

File: prometheus/exporter/docker_inspect.py (hunk 2 above)
Comment (fanyangCS) [reply prometheus/exporter/docker_inspect.py]: this is *not* unique, you need to combine it with task_role_name in order for the correct aggregation.

---
In reply to: [181279023](https://github.com/Microsoft/pai/pull/472#discussion_r181279023) [](ancestors = 181279023)

File: prometheus/exporter/job_exporter.py [hunk 6]
```diff
@@ -83,11 +88,13 @@ def genJobMetrics(logDir, gpuMetrics):
         outputFile.write(containerMemPerc)
 
 def main(argv):
+    logDir = argv[0]
+    timeSleep = int(argv[1])
+    iter = 0
     while(True):
-        logDir = argv[0]
-        timeSleep = int(argv[1])
+        print("job exporter running {0} iteration".format(str(iter)))
```
Comment (fanyangCS) [inline prometheus/exporter/job_exporter.py]: >iter [](start = 62, length = 4)

forget iter++?

Comment (YanjieGao) [reply prometheus/exporter/job_exporter.py]: add

**Pull request merged**